aiohttp/web_urldispatcher.py

```diff
@@ -1242,8 +1301,52 @@ def freeze(self) -> None:
         super().freeze()
         for resource in self._resources:
             resource.freeze()
+        if HAS_HYPERSCAN:
```

Comment:
Haven't tested this yet but I expect it will break downstream Home Assistant since there is code there to register new routes as new integrations get loaded after startup. It's pretty clear the intent here was to not allow that after freeze but that's currently the downstream use case and hard to change it.

Reply:
Good point.
We can rebuild hypedscan db in `UrlDispatcher.register_resource()` if the dispatcher was frozen.
I'm curious how Home Assistant works with route table on the fly.
If it adds new routes to started web server it should remove them at some point, isn't it?

Before moving forward, I'd like to add some benchmarks first.

Reply:
About to go to the airport, but will dig up the Home Assistant code this weekend. It predates me though so will take a bit to figure it out.

We don't have any benchmarks for the url dispatcher.  I'll look at downstream use cases and come up with some after I get back from travel next week.

Reply:
Have you safe flight!

Reply:
Once I figure out how the downstream code works for adding routes at run time. I'll see if I can make that into meaningful test cases here as well so we don't break anything.  AFAICT for a cursory review, routes are never removed when integrations are unloaded downstream.

Reply:
https://github.com/home-assistant/core/blob/445690588c855498b56561805ed652a12f36a21d/homeassistant/components/http/__init__.py#L620

Home Assistant effectively prevents freeze from happening and relies on the current implementation detail of route registration being implemented synchronously.

That's obviously not so great but it is currently effective. There are also a LOT of routes being added during and after startup as integrations get loaded.

Need to think about the best solution for downstream. 

Maybe the simplest solution is to allow routes to be added after freeze as long as freeze is called again as that should give some control
Of when the hyperdb gets rebuilt to reduce churning it. While it's still a breaking change for downstream, given freeze is being prevented that's seems reasonable to say downstream needs to adapt as it wasn't guaranteed behavior in the first place.  Maybe there is a better way though.   Need to think on it

Reply:
Maybe a new feature is to have a way to add routes after startup but don't guarantee they will be efficient until calling some method after new routes are added, which would internally trigger the hyperdb rebuild or any other data structures needed to be rebuilt. I think that would avoid the downstream monkey-patching and still fill the existing use case.

Reply:
> Maybe the simplest solution is to allow routes to be added after freeze as long as freeze is called again as that should give some control

I feel like any change to this should probably go into v4 only. We've currently always assumed that once the app has started, the routes are frozen. Other libraries depend on this behaviour in some cases, for example aiohttp-apischema evaluates the endpoint schema on startup from the routes and doesn't expect them to change at runtime.

If we do change this, we should have an explicit mechanism for other libraries to know about these changes. e.g. By registering callbacks which can be triggered when the routes are updated.

Reply:
BTW, is there a document for v4 roadmap?
What is the current status, blockers, things to be done?

Reply:
> > Maybe the simplest solution is to allow routes to be added after freeze as long as freeze is called again as that should give some control
> 
> I feel like any change to this should probably go into v4 only. We've currently always assumed that once the app has started, the routes are frozen. Other libraries depend on this behaviour in some cases, for example aiohttp-apischema evaluates the endpoint schema on startup from the routes and doesn't expect them to change at runtime.
> 
> If we do change this, we should have an explicit mechanism for other libraries to know about these changes. e.g. By registering callbacks which can be triggered when the routes are updated.

Maybe the safest way would be to require creating the app with `mutable_routes=True` and raise a warning if the user tries to modify routes after freeze without it telling them to carefully read the docs are there are performance considerations and rebuild must when changing view/routes after freeze.  That should ensure we don’t break any current behavior and provide a path forward for anything that is currently needing to modify after freeze.

Reply:
We literally just started thinking about this again, having triaged and sorted out a bunch of older issues.
Currently, it's just the v4 milestone, so we'll have to start getting through the tasks on there. I haven't checked whether the older things are still relevant, but have been adding things over time to the milestone that should be looked at before v4.

Reply:
I've been bucketing into the milestones with these conventions:

3.12: shorter term: https://github.com/aio-libs/aiohttp/milestone/35
4.0: old and longer term items: https://github.com/aio-libs/aiohttp/milestone/20

The caveat is that the 4.0 milestone curation is a bit stale.

Reply:
> The caveat is that the 4.0 milestone curation is a bit stale.

Yes, this is my fault. Sorry, guys.

Reply:
`mutable_routes=True` could be good idea, I should sleep on it.

Reply:
My criteria is just anything that is likely to result in breaking changes or deprecations I've added to the v4 milestone.

Reply:
> That should ensure we don’t break any current behavior and provide a path forward for anything that is currently needing to modify after freeze.

I feel like it's still breaking the libraries which aren't expected to advertise to users that they don't work when the flag is set.

Unless there's a clear tradeoff, I'd just put this into v4 without a flag. We just need to ensure that homeassistant has a way to continue hacking it in the mean time.

Reply:
We can introduce a new API for rebuilding, e.g.
```
with rebuild(app):
    app.router.add_get(...)
    app.add_subapp(...)
    app.middlewares.append(...)
```
The idea is *unfreezing* on `__enter__` and freezing back on `__exit__`.
The first version could work with route table only, middlewares and other things could be added iteratively on demand.

@bdraco could the proposed API fit Home Assistance usage scenario?

Reply:
> The idea is _unfreezing_ on `__enter__` and freezing back on `__exit__`.

That seems like a good way to avoid too much churn from recomputing everything after every call (if there are multiple calls making modifications).

Reply:
It also requires a locking probably to avoid concurrent route table modification from two web endpoints, for example.
Anyway, I want to finish optimizations first and maybe later return to freeze/unfreeze if we decide to use this approach.

Reply:
Freezing and unfreezing on enter and exit will likely be too many rebuilds.  ~297 routes get added at startup for Home Assistant. ~200 can be grouped together which would mean 200 rebuilds at startup

Reply:
Another method could be to keep adding to the fallback dicta new routes are added, clear the hyperscan db when a route is added after freeze, schedule a timer handle to rebuild the db 60 seconds later if one is not already scheduled

Reply:
> ~200 can be grouped together which would mean 200 rebuilds at startup

If they're grouped together, shouldn't that be 1 rebuild?

Reply:
Clarify: Very few can be grouped together. The result is 200 groups of various sizes, mostly 1

Reply:
Distribution of groupings looks something like 1, 9, 1, 7, 3, 3, 1, 1, 1, 1, 1 .....

However order is indeterminate as it relies on async task completion order over